### PR TITLE
Gear generator classes

### PR DESCRIPTION
**Pull request by Jojain**
I'm creating a new PR since I have rewrote most of the gear generator plugin (started here #13 ) with classes , becauseI think it is better to use and to work upon.

I have removed the gears that weren't mature enough, I may add them later as other PR but for now it will be easier to start with a few.

I've worked quite a lot on the maths of the bevel gears and they are now way more robust and flexible. However for the bevel gears and the spur gear, it happens that some set of parameters leads to non valid geometry. This happens mostly with helical gear. I don't have a set of parameter to reproduce the issue but one can try different value for the parameters and will enventually  encounter it but I will post one when I got one
I am quite confident with my modelization so I assume that on these cases the problem comes from OCP/OCCT

### CODE REVIEW

**Comment by Jojain** [comment]
An exemple of parameters that lead to wrong geometry, even though the parameters should yield a valid result.

```
from gear_generator import  Gear

m = 1.5
z = 22
b = 12
alpha = 14
helix_angle = 40

simple = Gear(m, z, b, alpha=alpha, raw = False, helix_angle=helix_angle).build()
```

image

**Comment by jmwright** [comment]
@Jojain Is this ready to be reviewed?

**Comment by Jojain** [comment]
@jmwright  yes !

File: tests/test_gear_generator.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+from plugins.gear_generator.gear_generator import *
+from unittest import TestCase
+from math import sin, radians
+
+class TestGearGenerator(TestCase):
+    def test_bevel_gear(self):
+        """
+        Tests if the bevel gear has been created successfully 
+        by checking if it's valid and it's resulting volume
+        Numerical values volumes were obtained with the Volume() functions on a Windows 10 machine with python 3.9.2
```
Comment (jmwright) [inline tests/test_gear_generator.py:11]: These tests are failing on Ubuntu 18.04. In two cases it looks like you can just reduce the number of decimal places to check. For this test, the following is what I see.
```
>       self.assertAlmostEqual(gear1.val().Volume(),2505.7531046471067)
E       AssertionError: 2505.7966225164664 != 2505.7531046471067 within 7 places (0.04351786935967539 difference)
```

File: tests/test_gear_generator.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        self.assertTrue(gear2.val().isValid())
+        self.assertAlmostEqual(gear1.val().Volume(),2505.7531046471067)
+
+
+    def test_bevel_gear_system(self):
+        """
+        Tests if the bevel gear system been created successfully 
+        by checking if it's valid and it's resulting volume
+        Numerical values volumes were obtained with the Volume() functions on a Windows 10 machine with python 3.9.2
```
Comment (jmwright) [inline tests/test_gear_generator.py:36]: Test result on Ubuntu 18.04.
```
>       self.assertAlmostEqual(pinion1.val().Volume(),483.3443067836134)
E       AssertionError: 483.3523144906429 != 483.3443067836134 within 7 places (0.008007707029548783 difference)
```

File: tests/test_gear_generator.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.assertTrue(pinion2.val().isValid())
+        self.assertTrue(gear2.val().isValid())
+        self.assertAlmostEqual(pinion2.val().Volume(),483.3523144907097)
+        self.assertAlmostEqual(gear2.val().Volume(),1126.605502545581)
+
+    def test_gear(self):
+        """
+        Tests if the gear has been created successfully 
+        by checking if it's valid and it's resulting volume
+        Numerical values volumes were obtained with the Volume() functions on a Windows 10 machine with python 3.9.2
```
Comment (jmwright) [inline tests/test_gear_generator.py:62]: Assert result on Ubuntu 18.04. This one is more concerning to me, like maybe invalid geometry is being generated on Linux.
```
>       self.assertAlmostEqual(gear1.val().Volume(),10113.473492717025)
E       AssertionError: 10045.87763052491 != 10113.473492717025 within 7 places (67.59586219211451 difference)
```

**Review by jmwright — COMMENTED** [review]
That last test (test_gear) I think is worth a closer look. If needed, I can get a version of CQ-editor running on your branch to see what the resulting gear looks like in Ubuntu 18.04.

File: plugins/gear_generator/README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+module = 2
+nb_teeth = 12
+width = 8
+gear = Gear(module, nb_teeth, width).build() #Instantiate a gear object and call it's build method to get the gear in a cq.Workplane
+```
+<img src="images/readme_example.PNG" width="300"/>
+
+Below is the list of implemented gear classes :
+```python
+Gear(args)
```
[the remark this answers is not in the document]
Comment (Jojain) [reply plugins/gear_generator/README.md:51]: In the introduction I presented some examples of the gear you can make but the dot list isn't making direct reference to the implemented classes.

Maybe each item list in the intro should be a class name and then add pictures of the kind of gear you can make with it (i.e straight or helical ) ?

If you find it confusing I can make this changes

[the remark this answers is not in the document]
Comment (jmwright) [reply plugins/gear_generator/README.md:51]: I think it's fine the way it is, I just wanted to check.

File: tests/test_gear_generator.py (hunk 3 above)
Comment (jmwright) [reply tests/test_gear_generator.py:62]: The resulting gear is a valid solid in Ubuntu 18.04. I even exported it to STL and sliced it to make sure. A discrepancy of 67.6 seems like more than rounding error to me, but it's not due to the shape being invalid.

Screenshot from 2021-04-30 13-11-02

Comment (Jojain) [reply tests/test_gear_generator.py:62]: That's 0.7 % difference, it seems quite high though. But if it end up looking as expected I don't know what we should do, should we not test the validity by volume ? I took this testing method from the cq tests, maybe you have idea of better testing ?

File: plugins/gear_generator/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ |  Base gear class
+ |  This class stores attributes that are shared by any types of gear
+ |  Other gear classes inherit from this class
+ |
+ |  Attributes :
+ |      m : gear modulus
+ |      b : gear tooth facewidth
+ |      z : gear number of teeth
+ |      p : gear pitch
+-- Suite  --
```
[the remark this answers is not in the document]
Comment (Jojain) [reply plugins/gear_generator/README.md:72]: I don't understand what you are talking about here ?

**Comment by Jojain** [comment]
There is a serious bug here in cq/OCCT, I went to see why the windows Gear test was failing and the result is that upon runs without any changes the result is either valid or not.
image
image

I just hit twice the render button in cq-editor and I got the above result. And it's not a graphical bug but a kernel bug since the obtained volumes are respectively : 10033.574314576623 for the valid shape and -6642.4418367933495 for the wrong one

File: plugins/gear_generator/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmwright) [reply plugins/gear_generator/README.md:72]: I think it's just a glitch with GitHub's rendering.

**Comment by Jojain** [comment]
I gave it another look and the problem seems rather complicated to identify. The problem seems to be most likely happening on helical gears but by trying enough I also got a non valid spur gear. 

If someone wants to look in depth in my implementation to try identify what's wrong that would be appreciate, otherwise I will let it like this for now, so  you can reject the PR.

I might look at it again and more in depth in the future but right now I prefer focusing on other things.

**Comment by adam-urbanczyk** [comment]
Maybe this is interesting/helpful for you https://github.com/meadiode/cq_gears ?

**Comment by Jojain** [comment]
I have seen that. I'm not planning to modify my implementation any time soon and he said that his were stable so if people needs gears I think we are better to lead them to his repo than mine.

That being said, (about the bevel gear at least) he created his gear by constructing faces and then assemble them into solid while I worked with 3D operation and the result is less stable.

It think its something to keep in memory that when you want to make something quite complex and very reusable it's probably better to use the face stiching approach

**Review by jmwright — APPROVED** [review]

**Comment by jmwright** [comment]
@Jojain I don't know that the conversation on this PR ever finished, but how would you feel about me merging this plugin? We could put a note at the top of the readme if it will be buggy for users.

**Comment by Jojain** [comment]
@jmwright Feel free to merge it. The issue encountered back then were from the kernel which I have no control over it. The issue might have disappeared with newer version of OCCT.

**Review by jmwright — APPROVED** [review]
Thanks @Jojain !

**Pull request merged**